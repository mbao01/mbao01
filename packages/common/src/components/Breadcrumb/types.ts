import { ReactNode } from "react";

export type BreadcrumbProps = React.HTMLAttributes<HTMLElement> & {
  separator?: ReactNode;
};

export type BreadcrumbListProps = React.HTMLAttributes<HTMLUListElement>;

export type BreadcrumbItemProps = React.HTMLAttributes<HTMLLIElement>;

export type BreadcrumbLinkProps = React.HTMLAttributes<HTMLAnchorElement> & {
  asChild?: boolean;
};

export type BreadcrumbPageProps = React.HTMLAttributes<HTMLSpanElement>;

export type BreadcrumbSeparatorProps = React.HTMLAttributes<HTMLLIElement>;

export type BreadcrumbEllipsisProps = React.HTMLAttributes<HTMLSpanElement>;
