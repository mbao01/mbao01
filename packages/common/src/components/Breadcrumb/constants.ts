import { cva } from "../../libs";

export const getBreadcrumbListClasses = cva(
  "flex flex-wrap items-center gap-1.5 break-words text-sm sm:gap-2.5"
);

export const getBreadcrumbItemClasses = cva("inline-flex items-center gap-1.5");

export const getBreadcrumbLinkClasses = cva("transition-colors");

export const getBreadcrumbPageClasses = cva("font-normal");

export const getBreadcrumbSeparatorClasses = cva("[&>svg]:w-3.5 [&>svg]:h-3.5");

export const getBreadcrumbEllipsisClasses = cva("flex h-9 w-9 items-center justify-center");
