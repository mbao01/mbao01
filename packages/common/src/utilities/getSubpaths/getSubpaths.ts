const PATH_SEPARATOR = "/";

const createPathObject = (
  segment: string,
  pathname: string,
  pathLabels?: Record<string, string>
) => ({
  href: { pathname },
  // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
  label: pathLabels?.[pathname] || segment, // always fallback to segment if label does not exist
});

export const getSubpaths = (
  pathname: string,
  pathLabels?: Record<string, string> | undefined,
  includeRoot?: boolean
) => {
  const segments = pathname?.split(PATH_SEPARATOR).filter(Boolean);

  const breadcrumbs = segments.map((segment, index) => {
    const path = `${PATH_SEPARATOR}${segments
      .slice(0, index + 1)
      .join(PATH_SEPARATOR)}`;

    return createPathObject(segment, path, pathLabels);
  });

  if (includeRoot) {
    const ROOT_PATH = PATH_SEPARATOR;
    breadcrumbs.unshift(createPathObject("", ROOT_PATH, pathLabels));
  }

  return breadcrumbs;
};
