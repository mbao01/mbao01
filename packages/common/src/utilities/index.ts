export * from "./theme";
export { cn } from "./cn";
export { getSubpaths } from "./getSubpaths";
