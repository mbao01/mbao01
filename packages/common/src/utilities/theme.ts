export type Theme = "dark" | "light";

export const getTheme = () => {
  if (typeof window === "undefined") return null;

  let t = window.localStorage.getItem("__theme") as Theme;
  if (!t)
    t = window.matchMedia("(prefers-color-scheme: dark)")?.matches
      ? "dark"
      : "light";
  return t;
};

export const saveTheme = (theme: Theme) => {
  document.body.setAttribute("data-theme", theme);
  window.localStorage.setItem("__theme", theme);
};
