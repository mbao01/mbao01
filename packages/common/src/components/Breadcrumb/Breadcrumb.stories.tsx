import type { Meta, StoryContext, StoryFn, StoryObj } from "@storybook/react";
import { getSubpaths } from "../../utilities";
import { Breadcrumb } from "./Breadcrumb";

type ArgsType = {
  pathname: string;
  subpathLabels?: Record<string, string>;
  includeRoot?: boolean;
};

const withBreadcrumb = (_: StoryFn, context: StoryContext<ArgsType>) => {
  const { pathname, subpathLabels, includeRoot } = context.args;
  const subpaths = getSubpaths(pathname, subpathLabels, includeRoot);

  return (
    <Breadcrumb>
      <Breadcrumb.List>
        {subpaths.map(({ href, label }, index) => (
          <>
            <Breadcrumb.Item key={href.pathname}>
              <Breadcrumb.Link asChild>
                <a href={href.pathname}>{label}</a>
              </Breadcrumb.Link>
            </Breadcrumb.Item>
            {subpaths.length > index + 1 ? <Breadcrumb.Separator /> : null}
          </>
        ))}
      </Breadcrumb.List>
    </Breadcrumb>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: "Components/Breadcrumb",
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  args: {
    pathname: "/profile/edit",
    subpathLabels: {
      "/": "Dashboard",
    },
    includeRoot: true,
  },
  argTypes: {
    pathname: {
      control: "text",
      description: "URL pathname",
    },
    subpathLabels: {
      control: "object",
      description: "Object mapping a subpath to a label",
    },
    includeRoot: {
      control: "boolean",
      description: "Include the root path or not",
    },
  },
  decorators: [withBreadcrumb],
} satisfies Meta<ArgsType>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: { pathname: "/" },
};

export const WithSubpathLabels: Story = {
  args: {
    subpathLabels: {
      "/": "Dashboard",
      "/profile": "Profile",
      "/profile/edit": "Edit Profile",
    },
  },
};

export const WithoutSubpathLabels: Story = {
  args: {
    pathname: "/profile/edit",
    includeRoot: false,
  },
};
