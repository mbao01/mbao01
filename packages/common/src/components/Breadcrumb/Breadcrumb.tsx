import { forwardRef } from "react";
import { Slot } from "@radix-ui/react-slot";
import { ChevronRightIcon, EllipsisIcon } from "lucide-react";
import type {
  BreadcrumbEllipsisProps,
  BreadcrumbItemProps,
  BreadcrumbLinkProps,
  BreadcrumbListProps,
  BreadcrumbPageProps,
  BreadcrumbProps,
  BreadcrumbSeparatorProps,
} from "./types";
import { cn } from "../../utilities";
import {
  getBreadcrumbEllipsisClasses,
  getBreadcrumbItemClasses,
  getBreadcrumbLinkClasses,
  getBreadcrumbListClasses,
  getBreadcrumbPageClasses,
  getBreadcrumbSeparatorClasses,
} from "./constants";

const Breadcrumb = ({ children, ...props }: BreadcrumbProps) => (
  <nav aria-label="breadcrumb" {...props}>
    {children}
  </nav>
);
Breadcrumb.displayName = "Breadcrumb";

const BreadcrumbList = forwardRef<HTMLOListElement, BreadcrumbListProps>(
  ({ className, ...props }, ref) => (
    <ol ref={ref} className={cn(getBreadcrumbListClasses(), className)} {...props} />
  )
);
BreadcrumbList.displayName = "BreadcrumbList";

const BreadcrumbItem = forwardRef<HTMLLIElement, BreadcrumbItemProps>(
  ({ className, ...props }, ref) => (
    <li ref={ref} className={cn(getBreadcrumbItemClasses(), className)} {...props} />
  )
);
BreadcrumbItem.displayName = "BreadcrumbItem";

const BreadcrumbLink = forwardRef<HTMLAnchorElement, BreadcrumbLinkProps>(
  ({ asChild, className, ...props }, ref) => {
    const Comp = asChild ? Slot : "a";

    return <Comp ref={ref} className={cn(getBreadcrumbLinkClasses(), className)} {...props} />;
  }
);
BreadcrumbLink.displayName = "BreadcrumbLink";

const BreadcrumbPage = forwardRef<HTMLSpanElement, BreadcrumbPageProps>(
  ({ className, ...props }, ref) => (
    <span
      ref={ref}
      role="link"
      aria-disabled="true"
      aria-current="page"
      className={cn(getBreadcrumbPageClasses(), className)}
      {...props}
    />
  )
);
BreadcrumbPage.displayName = "BreadcrumbPage";

const BreadcrumbSeparator = ({ children, className, ...props }: BreadcrumbSeparatorProps) => (
  <li
    role="presentation"
    aria-hidden="true"
    className={cn(getBreadcrumbSeparatorClasses(), className)}
    {...props}
  >
    {children ?? <ChevronRightIcon />}
  </li>
);
BreadcrumbSeparator.displayName = "BreadcrumbSeparator";

const BreadcrumbEllipsis = ({ className, ...props }: BreadcrumbEllipsisProps) => (
  <span
    role="presentation"
    aria-hidden="true"
    className={cn(getBreadcrumbEllipsisClasses(), className)}
    {...props}
  >
    <EllipsisIcon className="h-4 w-4" />
    <span className="sr-only">More</span>
  </span>
);
BreadcrumbEllipsis.displayName = "BreadcrumbEllipsis";

Breadcrumb.List = BreadcrumbList;
Breadcrumb.Item = BreadcrumbItem;
Breadcrumb.Link = BreadcrumbLink;
Breadcrumb.Page = BreadcrumbPage;
Breadcrumb.Separator = BreadcrumbSeparator;
Breadcrumb.Ellipsis = BreadcrumbEllipsis;

export { Breadcrumb };
