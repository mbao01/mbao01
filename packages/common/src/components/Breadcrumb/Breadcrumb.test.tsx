import { render, screen, within } from "@testing-library/react";
import { Slash } from "lucide-react";
import { getSubpaths } from "../../utilities/getSubpaths";
import { Breadcrumb } from "./Breadcrumb";

describe("Breadcrumb", () => {
  const renderBreadcrumb = (
    pathname: string,
    labels?: Record<string, string>,
    includeRoot?: boolean
  ) => {
    const subpaths = getSubpaths(pathname, labels, includeRoot);
    return render(
      <Breadcrumb>
        <Breadcrumb.List>
          {subpaths.map(({ href, label }, index) => (
            <>
              <Breadcrumb.Item key={href.pathname}>
                <Breadcrumb.Link asChild>
                  <a href={href.pathname}>{label}</a>
                </Breadcrumb.Link>
              </Breadcrumb.Item>
              {subpaths.length > index + 1 ? (
                <Breadcrumb.Separator>
                  <Slash />
                </Breadcrumb.Separator>
              ) : null}
            </>
          ))}
        </Breadcrumb.List>
      </Breadcrumb>
    );
  };

  it("renders empty list", () => {
    const { asFragment } = renderBreadcrumb("");

    expect(screen.getByRole("list")).toBeEmptyDOMElement();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders root path", () => {
    const { asFragment } = renderBreadcrumb("/", { "/": "Home" }, true);

    const listItemEl = within(screen.getByRole("listitem"));
    const anchorEl = listItemEl.getByRole("link", { name: "Home" });

    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", "/");
    expect(asFragment()).toMatchSnapshot();
  });

  it("shows unlabeled paths", () => {
    const { asFragment } = renderBreadcrumb("/profile/edit", { "/": "dashboard" }, true);

    [
      { name: "dashboard", href: "/" },
      { name: "profile", href: "/profile" },
      { name: "edit", href: "/profile/edit" },
    ].forEach(({ name, href }) => {
      const anchorEl = screen.getByRole("link", { name });

      expect(anchorEl).toBeInTheDocument();
      expect(anchorEl).toHaveAttribute("href", href);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("shows labeled paths", () => {
    const { asFragment } = renderBreadcrumb("/profile/edit", {
      "/profile": "My Profile",
      "/profile/edit": "Edit Profile",
    });

    [
      { name: "My Profile", href: "/profile" },
      { name: "Edit Profile", href: "/profile/edit" },
    ].forEach(({ name, href }) => {
      const anchorEl = screen.getByRole("link", { name });

      expect(anchorEl).toBeInTheDocument();
      expect(anchorEl).toHaveAttribute("href", href);
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
