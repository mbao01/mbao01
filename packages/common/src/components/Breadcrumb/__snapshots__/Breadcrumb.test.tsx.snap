// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Breadcrumb > renders empty list 1`] = `
<DocumentFragment>
  <nav
    aria-label="breadcrumb"
  >
    <ol
      class="flex flex-wrap items-center gap-1.5 break-words text-sm sm:gap-2.5"
    />
  </nav>
</DocumentFragment>
`;

exports[`Breadcrumb > renders root path 1`] = `
<DocumentFragment>
  <nav
    aria-label="breadcrumb"
  >
    <ol
      class="flex flex-wrap items-center gap-1.5 break-words text-sm sm:gap-2.5"
    >
      <li
        class="inline-flex items-center gap-1.5"
      >
        <a
          class="transition-colors"
          href="/"
        >
          Home
        </a>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;

exports[`Breadcrumb > shows labeled paths 1`] = `
<DocumentFragment>
  <nav
    aria-label="breadcrumb"
  >
    <ol
      class="flex flex-wrap items-center gap-1.5 break-words text-sm sm:gap-2.5"
    >
      <li
        class="inline-flex items-center gap-1.5"
      >
        <a
          class="transition-colors"
          href="/profile"
        >
          My Profile
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:w-3.5 [&>svg]:h-3.5"
        role="presentation"
      >
        <svg
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
      >
        <a
          class="transition-colors"
          href="/profile/edit"
        >
          Edit Profile
        </a>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;

exports[`Breadcrumb > shows unlabeled paths 1`] = `
<DocumentFragment>
  <nav
    aria-label="breadcrumb"
  >
    <ol
      class="flex flex-wrap items-center gap-1.5 break-words text-sm sm:gap-2.5"
    >
      <li
        class="inline-flex items-center gap-1.5"
      >
        <a
          class="transition-colors"
          href="/"
        >
          dashboard
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:w-3.5 [&>svg]:h-3.5"
        role="presentation"
      >
        <svg
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
      >
        <a
          class="transition-colors"
          href="/profile"
        >
          profile
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:w-3.5 [&>svg]:h-3.5"
        role="presentation"
      >
        <svg
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
      >
        <a
          class="transition-colors"
          href="/profile/edit"
        >
          edit
        </a>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;
